polyfill/index.d.ts

```diff
@@ -1637,3 +1637,79 @@ export namespace Temporal {
     export function timeZone(): Temporal.TimeZone;
   }
 }
+
+declare namespace IntlPolyfill {
+  type Formattable =
+    | Date
+    | Temporal.Instant
+    | Temporal.ZonedDateTime
+    | Temporal.PlainDate
+    | Temporal.PlainTime
+    | Temporal.PlainDateTime
+    | Temporal.PlainYearMonth
+    | Temporal.PlainMonthDay;
+
+  interface DateTimeFormatRangePart extends Intl.DateTimeFormatPart {
+    source: 'shared' | 'startRange' | 'endRange';
+  }
+
+  export interface DateTimeFormat extends Intl.DateTimeFormat {
+    /**
+     * Format a date into a string according to the locale and formatting
+     * options of this `Intl.DateTimeFormat` object.
+     *
+     * @param date The date to format.
+     */
+    format(date?: Formattable | number): string;
+
+    /**
+     * Allow locale-aware formatting of strings produced by
+     * `Intl.DateTimeFormat` formatters.
+     *
+     * @param date The date to format.
+     */
+    formatToParts(date?: Formattable | number): Intl.DateTimeFormatPart[];
+
+    /**
+     * Format a date range in the most concise way based on the locale and
+     * options provided when instantiating this `Intl.DateTimeFormat` object.
+     *
+     * @param startDate The start date of the range to format.
+     * @param endDate The start date of the range to format. Must be the same
+     * type as `startRange`.
+     */
+    formatRange<T extends Formattable>(startDate: T, endDate: T): string;
+    formatRange(startDate: Date | number, endDate: Date | number): string;
+
+    /**
+     * Allow locale-aware formatting of tokens representing each part of the
+     * formatted date range produced by `Intl.DateTimeFormat` formatters.
+     *
+     * @param startDate The start date of the range to format.
+     * @param endDate The start date of the range to format. Must be the same
+     * type as `startRange`.
+     */
+    formatRangeToParts<T extends Formattable>(startDate: T, endDate: T): DateTimeFormatRangePart[];
+    formatRangeToParts(startDate: Date | number, endDate: Date | number): DateTimeFormatRangePart[];
+  }
+
+  export const DateTimeFormat: {
+    /**
+     * Creates `Intl.DateTimeFormat` objects that enable language-sensitive
+     * date and time formatting.
+     */
+    new (locales?: string | string[], options?: Intl.DateTimeFormatOptions): DateTimeFormat;
+    (locales?: string | string[], options?: Intl.DateTimeFormatOptions): DateTimeFormat;
+
+    /**
+     * Get an array containing those of the provided locales that are supported
+     * in date and time formatting without having to fall back to the runtime's
+     * default locale.
+     */
+    supportedLocalesOf(locales: string | string[], options?: Intl.DateTimeFormatOptions): string[];
+  };
+}
+
+export { IntlPolyfill as Intl };
+
+export function toTemporalInstant(this: Date): Temporal.Instant;
```

Comment:
Is this syntax equivalent to using [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) to extend the `Date` interface?  If not, why is this syntax superior to using declaration merging here?

Reply:
No, this isn't declaration merging and doesn't affect the original Date object. It's used for `toTemporalInstance.call(myDate)`. My impression is the polyfill doesn't mutate date.

Reply:
The polyfill intentionally doesn't mutate Date or Intl by default, the reason is that users should opt in to whether they want their builtin objects modified or not. I do expect that many or most users of the polyfill will do so, so I'm not sure what's the best course of action for the TypeScript bindings here.

Reply:
The TypeScript bindings are designed to avoid mutating the native types, just like the polyfill. Declaration merging as Justin describes mutates the native types.